### PR TITLE
[SM6.10][HLK] LinAlg element access exec tests

### PR DESCRIPTION
**Pull request by V-FEXrt**
Adds HLK smoke test for LinAlg matrix element access

### CODE REVIEW

File: tools/clang/unittests/HLSLExec/LinAlgTests.cpp [hunk 1]
```diff
@@ -456,4 +464,187 @@ void DxilConf_SM610_LinAlg::SplatStore_Wave_I32() {
...
 
+static const char ElementAccessShader[] = R"(
+  RWByteAddressBuffer Input : register(u0);
+  RWByteAddressBuffer Output : register(u1);
+
+  // 0,0 = 0
+  // 0,1 = 4
+  // 1,0 = 16
+  // 3,3 = 60
+  // TODO: this assumes M=4,N=4
```
Comment (V-FEXrt) [inline tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: A lot of this test (pre and post copilot rewrite) makes a couple assumptions that won't really expand to full exec tests. 

Specifically, this test (implictly) assumes 4x4 matrix in a couple places and more generally the test framework assumes the following
- All individual elements read/written to the buffer are 4 bytes wide (float/int)
- All int results being read back for verification are <256 bytes (it only checks a single byte index of all 4)

The last point is the easiest to fix, but in general I think my stance is that we punt on fixing these until all the smoke tests are in that way we can iteratively solve the chicken-egg problem. If we make the tests any more complicated I feel we'll risk major bugs in the test

File: tools/clang/unittests/HLSLExec/LinAlgTests.cpp [hunk 2]
```diff
@@ -456,4 +464,187 @@ void DxilConf_SM610_LinAlg::SplatStore_Wave_I32() {
...
+  uint cordToByteOffset(uint2 coord) {
+    return (coord.x * 4 + coord.y) * 4;
```
Comment (V-FEXrt) [inline tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: 75% chance I picked the wrong X/Y order here. Getting indexes right is my kryptonite. Normally its fine because I can run a quick test to verify but in this case I can't actually run the code.

Comment (V-FEXrt) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: (Also I think RowMajor/ColMajor throws another wrench into everything)

File: tools/clang/unittests/HLSLExec/LinAlgTests.cpp (hunk 1 above)
Comment (tex3d) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: I think at minimum, we would want the smoke test to fit within required Tier 1 functionality, which might require different M & N dimensions. Implicitly requiring M=4,N=4 sounds like a bad starting assumption.

Comment (V-FEXrt) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: Yeah I'm not very excited about the 4x4 assumption at all and that should be fixed but I don't think we want quite as far as having tests for different M & N dims just yet.

My goal right now is to just exercise every operation with the most basic possible test. Since we can't actually run anything, every level of complexity on top of that is something that'll increase chances of a test bug in the code for our partners to discover during their bring up

Comment (tex3d) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: It’s not required to try variable dimensions, but if you choose 16x16, and put those in a #define or static constant, that would be fine.

Comment (V-FEXrt) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: I improved the structure of the test, and I'm a lot happier with that but I'm not realizing you may have been saying the minimum matrix size is 16x16 and we shouldn't even have 4x4 in the test?

Comment (V-FEXrt) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: hmm nope, I just checked the spec and the minimum is 4x4!

Comment (tex3d) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: I'm saying that 16x16 should be supported as part of required functionality, while 4x4 would be optional depending on what the device reports as supported (and we don't yet have that infrastructure finalized or implemented), so it's likely it won't be usable on real hardware, while 16x16 should be.

In other words, a smoke test shouldn't test a corner case that most real devices can't run.

Comment (tex3d) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: On that note also: I think F16 for all types would be within the required set, while F32 and I32 are not.

**Comment by V-FEXrt** [comment]
Attempting to fixing GH issues! will reopen in a moment

File: tools/clang/unittests/HLSLExec/LinAlgTests.cpp (hunk 1 above)
Comment (tex3d) [reply tools/clang/unittests/HLSLExec/LinAlgTests.cpp]: `16x16x16` with `F16` types will always be a supported combination within Tier 1 requirements, so I'm recommending we use that for the smoke test, rather than optional dimensions/component types.

**Comment by V-FEXrt** [comment]
https://github.com/microsoft/DirectXShaderCompiler/pull/8317 updated PR
